nav2_costmap_2d/test/integration/test_costmap_2d_publisher.cpp

```diff
@@ -0,0 +1,156 @@
+
+    callback_group_ = create_callback_group(
+      rclcpp::CallbackGroupType::MutuallyExclusive, false);
+
+    costmap_ = std::make_shared<nav2_costmap_2d::Costmap2D>(
+      100, 100, 0.05, 0, 0,
+      nav2_costmap_2d::LETHAL_OBSTACLE);
+
+    costmap_pub_ = std::make_shared<nav2_costmap_2d::Costmap2DPublisher>(
+      shared_from_this(),
+      costmap_.get(),
+      "map",
+      "dummy_costmap",
+      true);
+
+    rclcpp::SubscriptionOptions sub_option;
+    sub_option.callback_group = callback_group_;
+
+    std::string topic_name = "dummy_costmap_raw";
+    costmap_sub_ = this->create_subscription<nav2_msgs::msg::Costmap>(
+      topic_name,
+      rclcpp::QoS(rclcpp::KeepLast(1)).transient_local().reliable(),
+      std::bind(&TestCostmap2dPublisher::costmapCallback, this, std::placeholders::_1),
+      sub_option);
+
+    executor_ = std::make_shared<rclcpp::executors::SingleThreadedExecutor>();
+    executor_->add_callback_group(callback_group_, get_node_base_interface());
+    executor_thread_ = std::make_unique<nav2_util::NodeThread>(executor_);
+
+    return nav2_util::CallbackReturn::SUCCESS;
+  }
+
+  nav2_util::CallbackReturn
+  on_activate(const rclcpp_lifecycle::State &)
+  {
+    RCLCPP_INFO(get_logger(), "Activating");
+    costmap_pub_->on_activate();
+    return nav2_util::CallbackReturn::SUCCESS;
+  }
+
+  nav2_util::CallbackReturn
+  on_deactivate(const rclcpp_lifecycle::State &)
+  {
+    RCLCPP_INFO(get_logger(), "Deactivating");
+    return nav2_util::CallbackReturn::SUCCESS;
+  }
+
+  nav2_util::CallbackReturn
+  on_cleanup(const rclcpp_lifecycle::State &)
+  {
+    executor_thread_.reset();
+    return nav2_util::CallbackReturn::SUCCESS;
+  }
+
+  void publishCostmap()
+  {
+    costmap_pub_->publishCostmap();
+  }
+
+  void costmapCallback(const nav2_msgs::msg::Costmap::SharedPtr costmap)
+  {
+    promise_.set_value(costmap);
+  }
+
+  std::shared_ptr<nav2_costmap_2d::Costmap2DPublisher> costmap_pub_;
+  rclcpp::Subscription<nav2_msgs::msg::Costmap>::SharedPtr costmap_sub_;
+  std::shared_ptr<nav2_costmap_2d::Costmap2D> costmap_;
+
+  rclcpp::CallbackGroup::SharedPtr callback_group_;
+  rclcpp::executors::SingleThreadedExecutor::SharedPtr executor_;
+  std::unique_ptr<nav2_util::NodeThread> executor_thread_;
+  std::promise<nav2_msgs::msg::Costmap::SharedPtr> promise_;
+};
+
+class TestNode : public ::testing::Test
+{
+public:
+  TestNode()
+  {
+    costmap_publisher_ = std::make_shared<TestCostmap2dPublisher>("test_costmap_publisher");
+    costmap_publisher_->on_configure(costmap_publisher_->get_current_state());
+    costmap_publisher_->on_activate(costmap_publisher_->get_current_state());
+  }
+
+  ~TestNode()
+  {
+    costmap_publisher_->on_deactivate(costmap_publisher_->get_current_state());
+    costmap_publisher_->on_cleanup(costmap_publisher_->get_current_state());
+  }
+
+protected:
+  std::shared_ptr<TestCostmap2dPublisher> costmap_publisher_;
+};
+
+TEST_F(TestNode, costmap_pub_test)
```

Comment:
This doesn't also test the layers, which is what this PR adds in. I think you need to create a layer (or if this doesn't already have some default layers  to use like static - send it a bogus map on the map topic?) and check the value of the layer as well.

Reply:
> If we know that the costmap2d obj can be published then we also know all those inherited layers can be published.

I don't think that's necessarily true, someone could mess with the code and break the activation or publication of the individual layers separately from the costmap itself. The code blocks for the master layer's publication and the individual layers are separated in this PR so they're not sharing the same code.

```
  costmap_publisher_->on_activate();

  for (auto & layer_pub : layer_publishers_) {
    layer_pub->on_activate();
  }

...

     for (auto & layer_pub : layer_publishers_) {
        layer_pub->updateBounds(x0, xn, y0, yn);
      }

...


        for (auto & layer_pub : layer_publishers_) {
          layer_pub->publishCostmap();
        }
```

None of these lines have functional coverage to know if its working. What if you missed an `on_XYZ()` function? How would you know? The nature of this feature is to add publication of layers, so its important that publication of layers has coverage to know what you added actually accomplishes the task you set out to achieve.

Reply:
Its easy enough even without creating a new dummy plugin. If you have the costmap have a static map layer, you can publish a map and check that this map is reflected in the layer publication.